codeworld-game-server/src/CodeWorld/GameServer.hs

```diff
@@ -0,0 +1,267 @@
+
+type Games = HM.HashMap Key (MVar Game)
+
+data ServerState = ServerState
+    { games      :: MVar Games
+    , totalStats :: MVar TotalStats
+    , started    :: UTCTime
+    }
+
+
+-- Server state manipulation
+
+randomGameId :: IO GameId
+randomGameId = T.pack <$> replicateM 4 (randomRIO ('A', 'Z'))
+
+freshGame :: ServerState -> Int -> Signature -> IO (GameId, MVar Game)
+freshGame state playerCount sig = modifyMVar (games state) go
+  where
+    go games = do
+        gid <- randomGameId
+        if (sig, gid) `HM.member` games
+          then go games
+          else do
+            let game = Waiting playerCount (sig, gid) []
+            gameMV <- newMVar game
+            return (HM.insert (sig, gid) gameMV games, (gid, gameMV))
+
+
+joinGame :: WS.Connection -> MVar Game -> IO (Maybe PlayerId)
+joinGame conn gameMV = modifyMVar gameMV $ \game -> case game of
+        Waiting pc key plys | length plys < pc ->
+                let pid = length plys
+                    game' = Waiting pc key ((pid, conn) : plys)
+                in return (game', Just pid)
+        _ -> return (game, Nothing)
+
+tryStartGame :: MVar Game -> IO Bool
+tryStartGame gameMV = modifyMVar gameMV $ \game -> case game of
+        Waiting pc key plys | length plys == pc -> do
+                time <- getCurrentTime
+                return (Running time key plys, True)
+        _ -> return (game, False)
+
+getPlayers :: MVar Game -> IO [WS.Connection]
+getPlayers gameMVar = map snd . players <$> readMVar gameMVar
+
+getStats :: MVar Game -> IO (Int, Int)
+getStats gameMVar = go <$> readMVar gameMVar
+  where go (Waiting pc _ plys) = (length plys, pc)
+        go (Running t  _ plys) = (length plys, length plys)
+
+cleanup :: MVar Game -> PlayerId -> ServerState -> IO ()
+cleanup gameMV mypid state = do
+    done <- modifyMVar gameMV go
+    when done $ do
+        key <- gameKey <$> readMVar gameMV
+        modifyMVar_ (games state) (return . HM.delete key)
+  where
+    go g = let players' = filter ((/= mypid) . fst) (players g)
+           in return $ (g { players = players' }, null players')
+
+-- Communication
+
+sendServerMessage :: ServerMessage -> WS.Connection ->  IO ()
+sendServerMessage msg conn = WS.sendTextData conn (T.pack (show msg))
+
+getClientMessage :: WS.Connection -> IO ClientMessage
+getClientMessage conn = do
+    msg <- WS.receiveData conn
+    case readMaybe (T.unpack msg) of
+        Just msg -> return msg
+        Nothing -> fail "Invalid client message"
+
+broadcast :: ServerMessage -> MVar Game -> IO ()
+broadcast msg gameMV = withMVar gameMV $ \game ->
+    forM_ (players game) (sendServerMessage msg . snd)
+
+
+-- Statistics
+
+data CurrentStats = CurrentStats
+    { waitingGames :: !Int
+    , runningGames :: !Int
+    , connections  :: !Int
+    } deriving (Show, Generic)
+instance ToJSON CurrentStats
+
+data TotalStats = TotalStats
```

Comment:
Other possibly interesting numbers:

* Total messages sent
* Total messages received

Feel free to defer for later, though.